source/portable/NetworkInterface/TMS320F2838x/NetworkInterface.c

```diff
@@ -0,0 +1,304 @@
+    initInterfaceConfig.interruptNum[4] = INT_EMAC_RX1;
+
+    pInitCfg = Ethernet_initInterface(initInterfaceConfig);
+
+    Ethernet_getInitConfig(pInitCfg);
+
+    pInitCfg->dmaMode.InterruptMode = ETHERNET_DMA_MODE_INTM_MODE2;
+
+    //
+    // Assign the callbacks for Getting packet buffer when needed
+    // Releasing the TxPacketBuffer on Transmit interrupt callbacks
+    // Receive packet callback on Receive packet completion interrupt
+    //
+    pInitCfg->pfcbRxPacket = &Ethernet_receivePacketCallbackCustom;
+    pInitCfg->pfcbGetPacket = &Ethernet_getPacketBuffer;
+    pInitCfg->pfcbFreePacket = &Ethernet_releaseTxPacketBuffer;
+
+    //
+    //Assign the Buffer to be used by the Low level driver for receiving
+    //Packets. This should be accessible by the Ethernet DMA
+    //
+    pInitCfg->rxBuffer = Ethernet_rxBuffer;
+
+    //
+    // The Application handle is not used by this application
+    // Hence using a dummy value of 1
+    //
+    Ethernet_getHandle((Ethernet_Handle)1, pInitCfg , &emac_handle);
+
+    //
+    //Do global Interrupt Enable
+    //
+    (void)Interrupt_enableInProcessor();
+
+    //
+    //Assign default ISRs
+    //
+    Interrupt_registerHandler(INT_EMAC_TX0, Ethernet_transmitISR);
+    Interrupt_registerHandler(INT_EMAC_RX0, Ethernet_receiveISR);
+//    Interrupt_registerHandler(INT_EMAC, Ethernet_genericISR);
+
+    //
+    //Change the priority of the interrupt handlers
+    //
+
+    Interrupt_setPriority(INT_EMAC_TX0, 15);
+    Interrupt_setPriority(INT_EMAC_RX0, 15);
+//    Interrupt_setPriority(INT_EMAC, 15);
+
+    //
+    //Enable the default interrupt handlers
+    //
+    Interrupt_enable(INT_EMAC_TX0);
+    Interrupt_enable(INT_EMAC_RX0);
+//    Interrupt_enable(INT_EMAC);
+
+    if(pInitCfg == NULL){
+        xReturn = pdFAIL;
+    }else{
+        NetworkEndPoint_t* pxEndPoint = FreeRTOS_FirstEndPoint(pxInterface);
+        uint8_t MACAddress[ipMAC_ADDRESS_LENGTH_BYTES] = pxEndPoint->xMACAddress.ucBytes;
+        uint32_t MACAddressLow = (pxEndPoint->xMACAddress.ucBytes[0]) | (pxEndPoint->xMACAddress.ucBytes[1] << 8) | (pxEndPoint->xMACAddress.ucBytes[2] << 16) | (pxEndPoint->xMACAddress.ucBytes[3]<<24);
+        uint32_t MACAddressHigh = (pxEndPoint->xMACAddress.ucBytes[4]) | (pxEndPoint->xMACAddress.ucBytes[5] << 8);
+        Ethernet_setMACAddr(EMAC_BASE, 0, MACAddressHigh, MACAddressLow, ETHERNET_CHANNEL_0);
+        xReturn = pdPASS;
+    }
+
+    if(xEMACTaskHandle == NULL){
+        xTaskCreate(prvEMACDeferredInterruptHandlerTask, "EMACInt", configMINIMAL_STACK_SIZE, NULL, configMAX_PRIORITIES - 1, &xEMACTaskHandle);
+        configASSERT(xEMACTaskHandle);
+        pxMyInterface = pxInterface;
+    }
+
+    return xReturn;
+}
+
+static BaseType_t prvf28388d_NetworkInterfaceOutput(NetworkInterface_t* pxInterface, NetworkBufferDescriptor_t* const pxDescriptor, BaseType_t xReleaseAfterSend){
+    pxGlobalDescriptor = pxDescriptor;

```

Comment:
You are saving the packet buffer because `prvEMACDeferredInterruptHandlerTask()` will release it later on.
Please check `xReleaseAfterSend` : when false, you may not release the buffer.
And also, a few lines later in this function, you are releasing `pxDescriptor`, without invalidating the copy in `pxGlobalDescriptor`.